queries/stack-graphs.tsg

```diff
@@ -0,0 +1,1756 @@
+;; Also hook up `self`, `super`, and `crate` for declared modules.
+(mod_item name: (identifier) @name body: (declaration_list) @body) @mod {
+  edge @mod.def -> @body.def_values
+  edge @mod.def -> @body.def_types
+  edge @mod.def -> @body.def_any
+
+  ;; Define `self` as a node effectively identical to `@mod.def`. (Can't use
+  ;; the same node because TSG doesn't support aliases.)
+  node self
+  attr (self) source_node = @name, pop_symbol = "self", is_definition
+  edge @body.def_types -> self
+  edge self -> @body.def_values
+  edge self -> @body.def_types
+  edge self -> @body.def_any
+
+  ;; Bind `super` in a declared module as an alias for `self` in the enclosing
+  ;; module.
+  node super
+  attr (super) pop_symbol = "super"
+  node super_1
+  attr (super_1) push_symbol = "self"
+  node super_2
+  attr (super_2) push_symbol = "%types"
+  edge @body.def_types -> super
+  edge super -> super_1
+  edge super_1 -> super_2
+  edge super_2 -> @mod.scope
+}
+
+;; Trait members are not in scope for unqualified access, but they can be
+;; accessed by paths via the trait.
+(trait_item body: (declaration_list) @decls) @trait {
+  edge @trait.def -> @decls.def_values
+  edge @trait.def -> @decls.def_types
+}
+
+;; Install a path from a struct definition to its fields.
+(struct_item body: (field_declaration_list) @fields) @struct {
+  edge @struct.def -> @fields.fields
+}
+
+;; Note this is used by struct-like enum variants too.
+(field_declaration_list) @fields {
+  node @fields.fields
+  attr (@fields.fields) pop_symbol = "%fields"
+}
+
+(field_declaration_list (field_declaration name: (_) @field)) @fields {
+  node @field.def
+  attr (@field.def) node_definition = @field
+  edge @fields.fields -> @field.def
+}
+
+;; Install paths from an enum to its variants.
+(enum_item name: (type_identifier) @name ) @enum {
+  node @enum.variants_types
+  node t
+  attr (t) pop_symbol = "%types"
+  edge @enum.def -> t
+  edge t -> @enum.variants_types
+
+  node @enum.variants_values
+  node v
+  attr (v) pop_symbol = "%values"
+  edge @enum.def -> v
+  edge v -> @enum.variants_values
+
+  ;; The `%any` namespace supports looking up the identifier that's the last
+  ;; component of a use-declaration that imports from an enum.
+  node any
+  attr (any) pop_symbol = "%any"
+  node to_types
+  attr (to_types) push_symbol = "%types"
+  node to_values
+  attr (to_values) push_symbol = "%values"
+  edge @enum.def -> any
+  edge any -> to_types
+  edge to_types -> t
+  edge any -> to_values
+  edge to_values -> v
+}
+
+(enum_item
+  body: (enum_variant_list
+    (enum_variant
+      name: (identifier) @name
+      (field_declaration_list)? @fields
+    ) @variant
+  )
+) @enum {
+  node @variant.def
+  attr (@variant.def) node_definition = @name
+  if some @fields {
+    edge @enum.variants_types -> @variant.def
+    edge @variant.def -> @fields.fields
+  }
+  edge @enum.variants_values -> @variant.def
```

Comment:
I don't know if this will be a problem for how you set things up, but going through `%values` and `%types` guard nodes to the same definition can sometimes lead to value members becoming visible in type contexts and vice versa.

Reply:
I think it's okay to leave it like this for now, then.